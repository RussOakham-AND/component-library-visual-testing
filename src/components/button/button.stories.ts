import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction
const meta = {
  title: 'Components/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary: Story = {
  args: {
    children: 'Button',
    buttonType: 'primary',
    onPress: () => {
      window.console.log('Primary Button clicked!');
    },
  },
};

export const Secondary: Story = {
  args: {
    children: 'Button',
    buttonType: 'secondary',
    onPress: () => {
      window.console.log('Secondary Button clicked!');
    },
  },
};


export const Outline: Story = {
  args: {
    children: 'Button',
    buttonType: 'outline',
    onPress: () => {
      window.console.log('Outline Button clicked!');
    },
  },
};

export const ThemeOverride: Story = {
  args: {
    children: 'Button',
    buttonType: 'primary',
    theme: 2,
    onPress: () => {
      window.console.log('Outline Button clicked!');
    },
  },
};

