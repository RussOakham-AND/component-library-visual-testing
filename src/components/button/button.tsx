import { cva, type VariantProps } from 'class-variance-authority';
import { clsxMerge } from '@/common/utils/classNameUtils';
import { Button as AriaButton } from 'react-aria-components';

/**
 * Button styles for the Button component.>
 */
const buttonStyles = cva(
  'flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center',
  {
    variants: {
      buttonType: {
        primary:
          'bg-green-500 text-white border border-green-500 hover:bg-green-100',
        secondary:
          'bg-transparent text-green-500 border border-green-500 hover:bg-green-100 hover:opacity-100',
        warning:
          'bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600',
        link: 'text-black underline-offset-4 hover:underline',
        outline:
          'bg-white text-gray-600 hover:bg-gray-100 border hover:border-gray-100 border-gray-300 hover:shadow-md',
        disabled: 'bg-black text-white border-black cursor-not-allowed',
        error: 'bg-red-500 text-white border-red-500 hover:bg-red-600',
      },
      size: {
        default: ['text-base py-2 px-4'],
        small: ['text-sm py-1 px-2'],
        large: ['text-lg py-3 px-6'],
        xxl: ['text-2xl py-4 px-8'],
      },
      theme: {
        1: 'bg-red-500 text-white border border-red-500 hover:bg-red-100',
        2: 'bg-purple-500 rounded-md text-black border border-purple-500 hover:bg-violet-600',
        3: 'bg-blue-500 rounded-lg text-white border border-blue-500 hover:bg-blue-100',
      },
      rounded: {
        default: '',
        sm: 'rounded-sm',
        md: 'rounded-md',
        lg: 'rounded-lg',
        xl: 'rounded-xl',
        xxl: 'rounded-2xl',
        none: 'rounded-none',
        full: 'rounded-full',
      },
    },
    compoundVariants: [
      {
        buttonType: 'primary',
        size: 'default',
        rounded: 'default',
      },
    ],
    defaultVariants: {
      buttonType: 'primary',
      size: 'default',
      rounded: 'default',
    },
  }
);

type ButtonElementProps = React.ComponentProps<typeof AriaButton>;

interface ButtonProps
  extends ButtonElementProps,
    VariantProps<typeof buttonStyles> {}

function Button({
  className,
  buttonType,
  size,
  theme,
  rounded,
  children,
  ...props
}: ButtonProps) {
  return (
    <AriaButton
      className={clsxMerge(
        buttonStyles({ buttonType, size, theme, rounded }),
        className
      )}
      type="button"
      {...props}
    >
      {children}
    </AriaButton>
  );
}

export { Button, type ButtonElementProps, buttonStyles };
