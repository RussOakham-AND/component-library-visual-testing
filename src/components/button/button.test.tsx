import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Button } from './button';

describe('Button', () => {
  it('renders correctly', () => {
    const { container } = render(<Button>Click Me</Button>);

    expect(container.firstChild).toHaveClass('bg-green-500');
    expect(container.firstChild).toHaveClass('text-white');
    expect(container.firstChild).toHaveClass('px-4');
  });

  it('renders correctly with label', () => {
    const { container } = render(<Button>Click Me</Button>);

    expect(container).toHaveTextContent('Click Me');
  });

  it('render correctly with size prop', () => {
    const { container } = render(<Button size="default" />);

    const { container: containerSmall } = render(<Button size="small" />);

    const { container: containerLarge } = render(<Button size="large" />);

    const { container: containerXXL } = render(<Button size="xxl" />);

    expect(container.firstChild).toHaveClass('text-base');
    expect(containerSmall.firstChild).toHaveClass('text-sm');
    expect(containerLarge.firstChild).toHaveClass('text-lg');
    expect(containerXXL.firstChild).toHaveClass('text-2xl');
  });

  it('render correctly with rounded prop', () => {
    const { container } = render(<Button rounded="full" />);

    const { container: containerLg } = render(<Button rounded="lg" />);

    const { container: containerNone } = render(<Button rounded="none" />);

    const { container: containerXL } = render(<Button rounded="xl" />);

    const { container: containerXXL } = render(<Button rounded="xxl" />);

    const { container: containerDefault } = render(
      <Button rounded="default" />
    );

    const { container: containerSM } = render(<Button rounded="sm" />);

    expect(container).toMatchSnapshot();
    expect(containerLg).toMatchSnapshot();
    expect(containerNone).toMatchSnapshot();
    expect(containerXL).toMatchSnapshot();
    expect(containerXXL).toMatchSnapshot();
    expect(containerDefault).toMatchSnapshot();
    expect(containerSM).toMatchSnapshot();

    expect(container.firstChild).toHaveClass('rounded-full');
    expect(containerLg.firstChild).toHaveClass('rounded-lg');
    expect(containerNone.firstChild).toHaveClass('rounded-none');
    expect(containerXL.firstChild).toHaveClass('rounded-xl');
    expect(containerXXL.firstChild).toHaveClass('rounded-2xl');
    expect(containerSM.firstChild).toHaveClass('rounded-sm');
  });

  it('render correctly with buttonType prop', () => {
    const { container: containerPrimary } = render(
      <Button buttonType="primary" />
    );

    const { container } = render(<Button buttonType="secondary" />);

    const { container: containerWarning } = render(
      <Button buttonType="warning" />
    );

    const { container: containerOutline } = render(
      <Button buttonType="outline" />
    );

    const { container: containerDisabled } = render(
      <Button buttonType="disabled" />
    );

    const { container: containerError } = render(<Button buttonType="error" />);

    expect(container).toMatchSnapshot();
    expect(containerPrimary).toMatchSnapshot();
    expect(containerWarning).toMatchSnapshot();
    expect(containerOutline).toMatchSnapshot();
    expect(containerDisabled).toMatchSnapshot();
    expect(containerError).toMatchSnapshot();

    expect(containerPrimary.firstChild).toHaveClass(
      'flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4'
    );
    expect(container.firstChild).toHaveClass(
      'flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-transparent text-green-500 border border-green-500 hover:bg-green-100 hover:opacity-100 text-base py-2 px-4'
    );
    expect(containerOutline.firstChild).toHaveClass(
      'bg-white text-gray-600 hover:bg-gray-100 border hover:border-gray-100 border-gray-300 hover:shadow-md'
    );
    expect(containerWarning.firstChild).toHaveClass(
      'bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600'
    );

    expect(containerDisabled.firstChild).toHaveClass(
      'bg-black text-white border-black cursor-not-allowed'
    );

    expect(containerError.firstChild).toHaveClass(
      'bg-red-500 text-white border-red-500 hover:bg-red-600'
    );
  });
});
