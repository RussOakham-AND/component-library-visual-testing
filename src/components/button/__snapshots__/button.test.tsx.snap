// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > render correctly with buttonType prop 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-transparent text-green-500 border border-green-500 hover:bg-green-100 hover:opacity-100 text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with buttonType prop 2`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with buttonType prop 3`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600 text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with buttonType prop 4`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-white text-gray-600 hover:bg-gray-100 border hover:border-gray-100 border-gray-300 hover:shadow-md text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with buttonType prop 5`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-black text-white border-black cursor-not-allowed text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with buttonType prop 6`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-red-500 text-white border-red-500 hover:bg-red-600 text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-full"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 2`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-lg"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 3`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-none"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 4`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-xl"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 5`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-2xl"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 6`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4"
    data-rac=""
    type="button"
  />
</div>
`;

exports[`Button > render correctly with rounded prop 7`] = `
<div>
  <button
    class="flex flex-row gap-x-4 disabled:cursor-not-allowed items-center justify-center bg-green-500 text-white border border-green-500 hover:bg-green-hover text-base py-2 px-4 rounded-sm"
    data-rac=""
    type="button"
  />
</div>
`;
